File: vibeshield-python/src/schema_validator.py
import math
import re
from datetime import datetime
from typing import Any, Dict, List, Optional, Union

def is_valid_date(val: str) -> bool:
    if not isinstance(val, str):
        return False
    # Validate date structure: YYYY-MM-DD or YYYY-MM-DDTHH:MM:SS...
    if not re.match(r"^\d{4}-\d{2}-\d{2}(?:T\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:?\d{2})?)?$", val):
        return False
    for fmt in ('%Y-%m-%d', '%Y-%m-%dT%H:%M:%S', '%Y-%m-%dT%H:%M:%S.%fZ', '%Y-%m-%dT%H:%M:%SZ'):
        try:
            datetime.strptime(val, fmt)
            return True
        except ValueError:
            pass
    # Support fromisoformat for tz-aware strings
    try:
        datetime.fromisoformat(val.replace('Z', '+00:00'))
        return True
    except ValueError:
        pass
    return False

def validate_schema(data: Any, schema: Dict[str, Any]) -> Dict[str, Any]:
    errors = []
    if data is None or not isinstance(data, dict):
        return {
            "valid": False,
            "errors": [{"field": "", "message": "Root data must be a dictionary"}],
            "sanitized_data": None
        }

    sanitized_data = {}

    # Strict whitelist check at root level
    for key in data.keys():
        if key not in schema:
            errors.append({
                "field": key,
                "message": f'Unknown field "{key}" is not allowed',
                "value": data[key]
            })

    # Validate defined fields
    for key, field_schema in schema.items():
        value = data.get(key)
        # Check if key is actually in data (differentiating between None and missing)
        value_exists = key in data
        validate_field(
            value if value_exists else None,
            value_exists,
            field_schema,
            key,
            errors,
            sanitized_data,
            key
        )

    return {
        "valid": len(errors) == 0,
        "errors": errors,
        "sanitized_data": sanitized_data if len(errors) == 0 else None
    }

def validate_field(
    value: Any,
    exists: bool,
    field_schema: Dict[str, Any],
    path: str,
    errors: List[Dict[str, Any]],
    parent_sanitized: Dict[str, Any],
    key: str
) -> None:
    # Get configuration keys supporting both camelCase and snake_case
    required = field_schema.get("required", False)
    allow_null = field_schema.get("allowNull", field_schema.get("allow_null", False))
    trim = field_schema.get("trim", False)
    field_type = field_schema.get("type")
    field_min = field_schema.get("min")
    field_max = field_schema.get("max")
    format_rule = field_schema.get("format")
    pattern = field_schema.get("pattern")
    enum_values = field_schema.get("enum")
    sub_schema = field_schema.get("schema")
    element_schema = field_schema.get("elementSchema", field_schema.get("element_schema"))

    # 1. Required & Nullability checks
    if not exists:
        if required:
            errors.append({"field": path, "message": "Field is required", "expected": field_type})
        return

    if value is None:
        if allow_null:
            parent_sanitized[key] = None
        else:
            errors.append({"field": path, "message": "Value cannot be null", "value": value, "expected": field_type})
        return

    # 2. Strict Type Checking
    if field_type == "string":
        if not isinstance(value, str):
            errors.append({"field": path, "message": f"Expected type string, got {type(value).__name__}", "value": value, "expected": "string"})
            return
        if trim:
            value = value.strip()
    elif field_type == "number":
        # Reject bool (bool is subclass of int in python)
        if isinstance(value, bool) or not isinstance(value, (int, float)) or (isinstance(value, float) and math.isnan(value)):
            errors.append({"field": path, "message": f"Expected type number, got {type(value).__name__}", "value": value, "expected": "number"})
            return
    elif field_type == "boolean":
        if not isinstance(value, bool):
            errors.append({"field": path, "message": f"Expected type boolean, got {type(value).__name__}", "value": value, "expected": "boolean"})
            return
    elif field_type == "object":
        if not isinstance(value, dict):
            errors.append({"field": path, "message": f"Expected type object, got {type(value).__name__}", "value": value, "expected": "object"})
            return
    elif field_type == "array":
        if not isinstance(value, (list, tuple)):
            errors.append({"field": path, "message": f"Expected type array, got {type(value).__name__}", "value": value, "expected": "array"})
            return
        # Normalize to list
        value = list(value)

    # 3. Min/Max Checks
    if field_type == "string":
        if field_min is not None and len(value) < field_min:
            errors.append({"field": path, "message": f"Length must be at least {field_min}", "value": value})
        if field_max is not None and len(value) > field_max:
            errors.append({"field": path, "message": f"Length must be at most {field_max}", "value": value})
    elif field_type == "number":
        if field_min is not None and value < field_min:
            errors.append({"field": path, "message": f"Value must be at least {field_min}", "value": value})
        if field_max is not None and value > field_max:
            errors.append({"field": path, "message": f"Value must be at most {field_max}", "value": value})
    elif field_type == "array":
        if field_min is not None and len(value) < field_min:
            errors.append({"field": path, "message": f"Array must contain at least {field_min} items", "value": value})
        if field_max is not None and len(value) > field_max:
            errors.append({"field": path, "message": f"Array must contain at most {field_max} items", "value": value})

    # 4. Format Checks
    if field_type == "string" and format_rule:
        valid = True
        if format_rule == "email":
            valid = bool(re.match(r"^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$", value))
        elif format_rule == "uuid":
            valid = bool(re.match(r"^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$", value, re.IGNORECASE))
        elif format_rule == "url":
            valid = bool(re.match(r"^https?://[^\s/$.?#].[^\s]*$", value, re.IGNORECASE))
        elif format_rule == "ipv4":
            valid = bool(re.match(r"^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$", value))
        elif format_rule == "ipv6":
            valid = bool(re.match(r"^(?:[0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}$", value))
        elif format_rule == "phone":
            valid = bool(re.match(r"^\+?[1-9]\d{1,14}$", value))
        elif format_rule == "date":
            valid = is_valid_date(value)
        if not valid:
            errors.append({"field": path, "message": f"Invalid {format_rule} format", "value": value})

    # 5. Pattern Check
    if field_type == "string" and pattern:
        # pattern can be string or regex pattern object
        compiled = re.compile(pattern) if isinstance(pattern, str) else pattern
        if not compiled.search(value):
            # Try full match or search
            if not compiled.match(value):
                errors.append({"field": path, "message": f"Value does not match pattern: {pattern}", "value": value})

    # 6. Enum Check
    if enum_values is not None:
        if value not in enum_values:
            errors.append({"field": path, "message": f"Value must be one of: {enum_values}", "value": value})

    # 7. Recursive Object Validation
    if field_type == "object":
        if sub_schema:
            sub_sanitized = {}
            for sub_key in value.keys():
                if sub_key not in sub_schema:
                    errors.append({
                        "field": f"{path}.{sub_key}",
                        "message": f'Unknown field "{sub_key}" is not allowed',
                        "value": value[sub_key]
                    })
            for sub_key, sub_field_schema in sub_schema.items():
                sub_value_exists = sub_key in value
                validate_field(
                    value.get(sub_key) if sub_value_exists else None,
                    sub_value_exists,
                    sub_field_schema,
                    f"{path}.{sub_key}",
                    errors,
                    sub_sanitized,
                    sub_key
                )
            parent_sanitized[key] = sub_sanitized
        else:
            parent_sanitized[key] = dict(value)
        return

    # 8. Recursive Array Element Validation
    if field_type == "array":
        if element_schema:
            sub_sanitized = []
            for i, element_value in enumerate(value):
                element_path = f"{path}[{i}]"
                temp_parent = {}
                validate_field(element_value, True, element_schema, element_path, errors, temp_parent, "element")
                if "element" in temp_parent:
                    sub_sanitized.append(temp_parent["element"])
            parent_sanitized[key] = sub_sanitized
        else:
            parent_sanitized[key] = list(value)
        return

    parent_sanitized[key] = value

File: vibeshield-ts/src/core/schema-validator.ts
export interface ValidationError {
  field: string;
  message: string;
  value?: any;
  expected?: string;
}

export interface ValidationResult {
  valid: boolean;
  errors: ValidationError[];
  sanitizedData?: any;
}

export interface SchemaDefinition {
  [fieldName: string]: FieldSchema;
}

export interface FieldSchema {
  type: 'string' | 'number' | 'boolean' | 'object' | 'array';
  required?: boolean;
  format?: 'email' | 'uuid' | 'url' | 'date' | 'phone' | 'ipv4' | 'ipv6';
  min?: number;
  max?: number;
  pattern?: RegExp;
  enum?: any[];
  schema?: SchemaDefinition;
  elementSchema?: FieldSchema;
  allowNull?: boolean;
  trim?: boolean;
}

export function validateSchema(data: unknown, schema: SchemaDefinition): ValidationResult {
  const errors: ValidationError[] = [];

  if (data === null || data === undefined || typeof data !== 'object' || Array.isArray(data)) {
    return {
      valid: false,
      errors: [{ field: '', message: 'Root data must be an object' }]
    };
  }

  const sanitizedData: any = {};

  // Strict whitelist check for unknown fields at root level
  const dataKeys = Object.keys(data as any);
  for (const key of dataKeys) {
    if (!(key in schema)) {
      errors.push({
        field: key,
        message: `Unknown field "${key}" is not allowed`,
        value: (data as any)[key]
      });
    }
  }

  // Validate expected fields
  for (const key of Object.keys(schema)) {
    const fieldSchema = schema[key];
    const value = (data as any)[key];
    validateField(value, fieldSchema, key, errors, sanitizedData, key);
  }

  return {
    valid: errors.length === 0,
    errors,
    sanitizedData: errors.length === 0 ? sanitizedData : undefined
  };
}

function validateField(
  value: any,
  fieldSchema: FieldSchema,
  path: string,
  errors: ValidationError[],
  parentSanitized: any,
  key: string
): void {
  // 1. Required & Nullability Checks
  if (value === undefined) {
    if (fieldSchema.required) {
      errors.push({ field: path, message: 'Field is required', expected: fieldSchema.type });
    }
    return;
  }

  if (value === null) {
    if (fieldSchema.allowNull) {
      parentSanitized[key] = null;
    } else {
      errors.push({ field: path, message: 'Value cannot be null', value, expected: fieldSchema.type });
    }
    return;
  }

  const { type } = fieldSchema;

  // 2. Strict Type Checking (no type coercion)
  if (type === 'string') {
    if (typeof value !== 'string') {
      errors.push({ field: path, message: `Expected type string, got ${typeof value}`, value, expected: 'string' });
      return;
    }
    if (fieldSchema.trim) {
      value = value.trim();
    }
  } else if (type === 'number') {
    if (typeof value !== 'number' || Number.isNaN(value)) {
      errors.push({ field: path, message: `Expected type number, got ${typeof value}`, value, expected: 'number' });
      return;
    }
  } else if (type === 'boolean') {
    if (typeof value !== 'boolean') {
      errors.push({ field: path, message: `Expected type boolean, got ${typeof value}`, value, expected: 'boolean' });
      return;
    }
  } else if (type === 'object') {
    if (typeof value !== 'object' || Array.isArray(value) || value === null) {
      errors.push({ field: path, message: `Expected type object, got ${Array.isArray(value) ? 'array' : typeof value}`, value, expected: 'object' });
      return;
    }
  } else if (type === 'array') {
    if (!Array.isArray(value)) {
      errors.push({ field: path, message: `Expected type array, got ${typeof value}`, value, expected: 'array' });
      return;
    }
  }

  // 3. Min/Max Checks
  if (type === 'string') {
    if (fieldSchema.min !== undefined && value.length < fieldSchema.min) {
      errors.push({ field: path, message: `Length must be at least ${fieldSchema.min}`, value });
    }
    if (fieldSchema.max !== undefined && value.length > fieldSchema.max) {
      errors.push({ field: path, message: `Length must be at most ${fieldSchema.max}`, value });
    }
  } else if (type === 'number') {
    if (fieldSchema.min !== undefined && value < fieldSchema.min) {
      errors.push({ field: path, message: `Value must be at least ${fieldSchema.min}`, value });
    }
    if (fieldSchema.max !== undefined && value > fieldSchema.max) {
      errors.push({ field: path, message: `Value must be at most ${fieldSchema.max}`, value });
    }
  } else if (type === 'array') {
    if (fieldSchema.min !== undefined && value.length < fieldSchema.min) {
      errors.push({ field: path, message: `Array must contain at least ${fieldSchema.min} items`, value });
    }
    if (fieldSchema.max !== undefined && value.length > fieldSchema.max) {
      errors.push({ field: path, message: `Array must contain at most ${fieldSchema.max} items`, value });
    }
  }

  // 4. Format Checks
  if (type === 'string' && fieldSchema.format) {
    let valid = true;
    if (fieldSchema.format === 'email') {
      valid = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/.test(value);
    } else if (fieldSchema.format === 'uuid') {
      valid = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i.test(value);
    } else if (fieldSchema.format === 'url') {
      valid = /^https?:\/\/[^\s/$.?#].[^\s]*$/i.test(value);
    } else if (fieldSchema.format === 'ipv4') {
      valid = /^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/.test(value);
    } else if (fieldSchema.format === 'ipv6') {
      valid = /^(?:[0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}$/.test(value);
    } else if (fieldSchema.format === 'phone') {
      valid = /^\+?[1-9]\d{1,14}$/.test(value);
    } else if (fieldSchema.format === 'date') {
      valid = !isNaN(Date.parse(value)) && /^\d{4}-\d{2}-\d{2}(?:T\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:?\d{2})?)?$/.test(value);
    }
    if (!valid) {
      errors.push({ field: path, message: `Invalid ${fieldSchema.format} format`, value });
    }
  }

  // 5. Pattern Check
  if (type === 'string' && fieldSchema.pattern) {
    if (!fieldSchema.pattern.test(value)) {
      errors.push({ field: path, message: `Value does not match pattern: ${fieldSchema.pattern.toString()}`, value });
    }
  }

  // 6. Enum Check
  if (fieldSchema.enum) {
    if (!fieldSchema.enum.includes(value)) {
      errors.push({ field: path, message: `Value must be one of: [${fieldSchema.enum.join(', ')}]`, value });
    }
  }

  // 7. Recursive Object Validation
  if (type === 'object') {
    if (fieldSchema.schema) {
      const subSanitized: any = {};
      const subKeys = Object.keys(value);
      for (const subKey of subKeys) {
        if (!(subKey in fieldSchema.schema)) {
          errors.push({
            field: `${path}.${subKey}`,
            message: `Unknown field "${subKey}" is not allowed`,
            value: value[subKey]
          });
        }
      }

      for (const subKey of Object.keys(fieldSchema.schema)) {
        const subFieldSchema = fieldSchema.schema[subKey];
        const subValue = value[subKey];
        validateField(subValue, subFieldSchema, `${path}.${subKey}`, errors, subSanitized, subKey);
      }
      parentSanitized[key] = subSanitized;
    } else {
      parentSanitized[key] = { ...value };
    }
    return;
  }

  // 8. Recursive Array Element Validation
  if (type === 'array') {
    if (fieldSchema.elementSchema) {
      const subSanitized: any[] = [];
      for (let i = 0; i < value.length; i++) {
        const elementPath = `${path}[${i}]`;
        const elementValue = value[i];
        const tempParent: any = {};
        validateField(elementValue, fieldSchema.elementSchema, elementPath, errors, tempParent, 'element');
        subSanitized.push(tempParent.element);
      }
      parentSanitized[key] = subSanitized;
    } else {
      parentSanitized[key] = [...value];
    }
    return;
  }

  parentSanitized[key] = value;
}
